### PR TITLE
Colorbar enhancements

### PR DESCRIPTION
**Pull request by isentropic**
Minimal working colorbar enhancements, pyplot only. Response to https://github.com/JuliaPlots/Plots.jl/issues/3174
Colorbars  need some :heart: too )

```julia
heatmap(rand(10,10), colorbar_ticks=[0.3,0.6])
```
image

### CODE REVIEW

File: src/args.jl [hunk 1]
```diff
@@ -151,54 +151,54 @@ const _shape_keys = Symbol[
 
 const _allMarkers = vcat(:none, :auto, _shape_keys) #sort(collect(keys(_shapes))))
 const _markerAliases = Dict{Symbol,Symbol}(
```
Comment (isentropic) [inline src/args.jl:153]: Aligning the dict properly

File: src/args.jl [hunk 2]
```diff
@@ -363,104 +363,115 @@ const _plot_defaults = KW(
...
+    :projection                 => :none,             # can also be :polar or :3d
+    :aspect_ratio               => :auto,             # choose from :none or :equal
+    :margin                     => 1mm,
+    :left_margin                => :match,
+    :top_margin                 => :match,
+    :right_margin               => :match,
+    :bottom_margin              => :match,
+    :subplot_index              => -1,
+    :colorbar_title             => "",
+    :colorbar_titlefontsize     => 10,
```
Comment (isentropic) [inline src/args.jl:419]: New colorbar args

**Comment by isentropic** [comment]
Most importantly I need some comments about the Plots backend restructure and creation of the colobars.jl file.
currently only colorbar_ticks implemented, but the rest of the arguments should be easy to implement in the backend code only. Just want to make sure that the Plots pipeline altering is done fine here.
This does not break any previous behavior, like usual it would be up to the backend to support the these colorbar arguments. `colorbar_ticks=:native`should default back the original behavior for all backends.

File: src/colorbars.jl [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        cbar_gradient
+    else
+        nothing
+    end
+end
+
+hascolorbar(series::Series) = colorbar_style(series) !== nothing
+hascolorbar(sp::Subplot) = sp[:colorbar] != :none && any(hascolorbar(s) for s in series_list(sp))
+
+function optimal_colorbar_ticks_and_labels(sp::Subplot, ticks = nothing)
```
Comment (daschw) [inline src/colorbars.jl:74]: Do we really need to duplicate this code from axes.jl?

Comment (isentropic) [reply src/colorbars.jl:74]: the reason for duplication is that ive decided that colorbar is a subplot arg, hence it does not have an axis (x, y, or z)

Comment (isentropic) [reply src/colorbars.jl:74]: axis is not defined, hence the colorbar args need different treatment

Comment (isentropic) [reply src/colorbars.jl:74]: perhaps it could just overload the method, but is it better?

Comment (daschw) [reply src/colorbars.jl:74]: maybe we should have the main logic of `optimal_ticks_and_labels` in axes.jl in a method
```julia
optimal_ticks_and_labels(amin, amax, scale, formatter, ticks)
```
Then we can define the existing version as
```julia
optimal_ticks_and_labels(sp::Subplot, axis::Axis, ticks = nothing) =
    optimal_ticks_and_labels(axis_limits(sp, axis[:letter])..., axis[:scale], axis[:formatter], ticks)
```
and use the first version also for colorbars. I think it's easier to maintain if we do not have duplicate code.

Comment (isentropic) [reply src/colorbars.jl:74]: yes this sounds much better

File: src/colorbars.jl [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    #
+    # If the axis input was a Date or DateTime use a special logic to find
+    # "round" Date(Time)s as ticks
+    # This bypasses the rest of optimal_ticks_and_labels, because
+    # optimize_datetime_ticks returns ticks AND labels: the label format (Date
+    # or DateTime) is chosen based on the time span between amin and amax
+    # rather than on the input format
+    # TODO: maybe: non-trivial scale (:ln, :log2, :log10) for date/datetime
+    if ticks === nothing && scale == :identity
+        if sp[:colorbar_formatter] == RecipesPipeline.dateformatter
```
Comment (isentropic) [inline src/colorbars.jl:91]: what about here? sp[:colorbar formatter] is not part of axis

File: src/colorbars.jl [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    end
+
+    # @show unscaled_ticks labels
+    # labels = Showoff.showoff(unscaled_ticks, scale == :log10 ? :scientific : :auto)
+    unscaled_ticks, labels
+end
+
+# return (continuous_values, discrete_values) for the ticks on this axis
+function get_colorbar_ticks(sp::Subplot; update = true)
+    if update || !haskey(sp.attr, :colorbar_optimized_ticks)
```
Comment (isentropic) [inline src/colorbars.jl:159]: Also this case

File: src/colorbars.jl (hunk 3 above)
Comment (daschw) [reply src/colorbars.jl:74]: I made a PR for this on the colorbar_enhancements branch of your fork.

**Review by isentropic — COMMENTED** [review]
Done

File: src/args.jl [hunk 6]
```diff
@@ -363,104 +363,115 @@ const _plot_defaults = KW(
...
+    :colorbar_title_location    => :center,           # also :left or :right
+    :colorbar_fontfamily        => :match,
```
Comment (isentropic) [inline src/args.jl:421]: Where do I do the matching the colorbar_fontfamily, colorbartitlefont and colorbar_titlefontcolor to the rest of the fonts?

**Review by isentropic — COMMENTED** [review]
I cannot find the colorbar font matching (setting sp[:colorbarfontcolor] =:black for example?). Currently all of the items have :match but I could not find where the calculation is made. 

Other than that, now the colorbarscale works too, with the PyPlot backend

**Comment by isentropic** [comment]
I checked recipespipeline, but cannot find it there too

File: src/args.jl (hunk 6 above)
Comment (daschw) [reply src/args.jl:421]: I think you are looking for https://github.com/JuliaPlots/Plots.jl/blob/f76390c20d188a35f517a8cf467c061325d1c22e/src/args.jl#L1342-L1344

**Comment by isentropic** [comment]
Nvrmind, i got it figured out, the matching happens in the `args.jl`. I'd say it is working pretty good now

```julia
plot = histogram2d(rand(1000,1000), colorbar_tickfontfamily="Serif", colorbar_tickfontcolor="Red", colorbar_scale=:log10)
```
image

**Comment by isentropic** [comment]
@daschw could you perhaps follow along with gr implementation for this? 

@BeastyBlacksmith pgfplots could also benefit too. 

I believe that most of the pipeline processing has been accomplished, there should not be a lot backend code to be implemented. If anything, it should make construction of the colorbar easier as relevant kws are introduced.

We also need to think about additional kws like: colorbar width, height, tick direction, extend and what not

**Comment by isentropic** [comment]
A lot of ideas were mentioned in #3174, but I believe we can't blindly accept all requests and should decide on the reasonable scope to cover. I personally think that we should cover as many as the backends can easily handle. 

Perhaps even run a poll for the needed args and decide on the scope.

**Comment by mkborregaard** [comment]
> I personally think that we should cover as many as the backends can easily handle.

Sounds like a good pragmatic approach to me

**Pull request merged**